packages/analytics-js/src/components/pluginsManager/PluginsManager.ts

```diff
@@ -94,101 +96,117 @@ class PluginsManager implements IPluginsManager {
+        configurationStatusStr: 'Storage encryption is enabled',
+        activePluginName: state.storage.encryptionPluginName.value,
+        supportedPlugins: Object.values(StorageEncryptionVersionsToPluginNameMap),
+      },
+      {
+        configurationStatus: () => state.storage.migrate.value,
+        configurationStatusStr: 'Storage migration is enabled',
+        supportedPlugins: ['StorageMigrator'],
+      },
+    ];
+
+    const addMissingPlugins = false;
+    pluginGroupsToProcess.forEach(group => {
+      if (group.configurationStatus()) {
+        pluginsToLoadFromConfig = pluginsToLoadFromConfig.filter(
+          group.activePluginName
+            ? pluginName =>
+                !(
+                  pluginName !== group.activePluginName &&
+                  group.supportedPlugins.includes(pluginName)
+                )
+            : pluginName => isDefined(pluginName), // pass through
+        );
+
+        this.addMissingPlugins(group, addMissingPlugins, pluginsToLoadFromConfig);
+      } else {
+        pluginsToLoadFromConfig = pluginsToLoadFromConfig.filter(
+          group.basePlugins
+            ? pluginName =>
+                !(
+                  group.basePlugins?.includes(pluginName) ||
+                  group.supportedPlugins.includes(pluginName)
+                )
+            : pluginName => !group.supportedPlugins.includes(pluginName),
+        );
       }
+    });
 
-      pluginsToLoadFromConfig = pluginsToLoadFromConfig.filter(
-        pluginName => pluginName !== 'BeaconQueue',
-      );
-    }
-
-    // Enforce default cloud mode event delivery queue plugin is none exists
-    if (
-      !pluginsToLoadFromConfig.includes('XhrQueue') &&
-      !pluginsToLoadFromConfig.includes('BeaconQueue')
-    ) {
-      pluginsToLoadFromConfig.push('XhrQueue');
-    }
-
-    // Device mode destinations related plugins
-    if (
-      getNonCloudDestinations(state.nativeDestinations.configuredDestinations.value ?? [])
-        .length === 0
-    ) {
-      pluginsToLoadFromConfig = pluginsToLoadFromConfig.filter(
-        pluginName =>
-          ![
-            'DeviceModeDestinations',
-            'DeviceModeTransformation',
-            'NativeDestinationQueue',
-          ].includes(pluginName),
-      );
-    }
-
-    // Consent Management related plugins
-    const supportedConsentManagerPlugins: string[] = Object.values(ConsentManagersToPluginNameMap);
-    let filterCondition = (pluginName: PluginName) =>
-      !(
-        pluginName !== state.consents.activeConsentManagerPluginName.value &&
-        supportedConsentManagerPlugins.includes(pluginName)
-      );
+    return [...(Object.keys(getMandatoryPluginsMap()) as PluginName[]), ...pluginsToLoadFromConfig];
+  }
 
-    if (!state.consents.enabled.value) {
-      filterCondition = (pluginName: PluginName) =>
-        !supportedConsentManagerPlugins.includes(pluginName);
+  private addMissingPlugins(
```

Comment:
Are we adding missing plugins back to the plugin list if, by mistake, they are not provided in the load options?

Reply:
That's the future goal, but we'll only warn them at this point.
However, the current design is such a way that moving to that phase will be pretty easy.